src/Storages/S3Queue/S3QueueSource.cpp

```diff
@@ -85,19 +112,44 @@ StorageS3QueueSource::ObjectInfoPtr StorageS3QueueSource::FileIterator::nextImpl
     return {};
 }
 
-std::pair<StorageS3QueueSource::ObjectInfoPtr, S3QueueOrderedFileMetadata::BucketInfoPtr>
+void StorageS3QueueSource::FileIterator::returnForRetry(Source::ObjectInfoPtr object_info)
+{
+    chassert(object_info);
+    if (metadata->useBucketsForProcessing())
+    {
+        const auto bucket = metadata->getBucketForPath(object_info->relative_path);
+        listed_keys_cache[bucket].keys.emplace_front(object_info);
+    }
+    else
+    {
+        objects_to_retry.push_back(object_info);
```

Comment:
Why is a key placed at the front of the bucketed cache? Is this to prioritize the failed keys over the keys that are not yet read? Can the same key use up all the `maxRetries` quota if it's placed at the front over and over again?

Reply:
We use buckets when we have `Ordered` mode of `S3Queue` and if one of the following conditions is true: "local_per_server_processing_threads" > 1 (`s3queue_processing_threads_num` setting) or "distributed_logical_threads" > 1 (`s3queue_buckets` setting). So because it is `Ordered` mode - order of processing files is important - we must always process them in alphanumeric order. If we have buckets - we need to maintain order only in scope of buckets. E.g. for each bucket there is only 1 thread which processes this bucket and it processes all files inside the bucket strictly in order (but processing among different buckets is parallelized). That is why we do `push_front` here - because this `object_info` here is ordered less than all remaining objects in `listed_keys_cache[bucket].keys`. But in line `objects_to_retry.push_back(object_info);` we do `push_back`, because 
(1) if it is Ordered mode - there will be only 1 thread which processing everything (as `metadata->useBucketsForProcessing() == 0` in this code branch), so in fact `objects_to_retry` must always have size == 1 in this case (I do need to add assertion for this!).
(2) if it is Unordered mode - in this case we do not care about order at all, but as here we can have any number of processing threads, `objects_to_retry.size() > 1` unlike case (1).

Reply:
Thanks for the detailed explanation!